src/Features/Core/Portable/EditAndContinue/EditSession.cs

```diff
@@ -261,38 +261,82 @@ private async Task<ActiveStatementsMap> GetBaseActiveStatementsAsync(Cancellatio
         {
-            changedOrAddedDocuments.Clear();
+            var newDocumentIds = newSolution.GetDocumentIdsWithFilePath(sourceFilePath);
+            if (newDocumentIds.IsEmpty)
+            {
+                // file does not exist or has been removed:
+                return !oldSolution.GetDocumentIdsWithFilePath(sourceFilePath).IsEmpty;
+            }
 
-            if (!newProject.SupportsEditAndContinue())
+            // it suffices to check the content of one of the document if there are multiple linked ones:
+            var documentId = newDocumentIds.First();
+            var oldDocument = oldSolution.GetDocument(documentId);
+            if (oldDocument == null)
             {
-                return;
+                // file has been added
+                return true;
             }
 
-            var oldProject = oldSolution.GetProject(newProject.Id);
+            var newDocument = newSolution.GetRequiredDocument(documentId);
+            return oldDocument != newDocument && !await ContentEqualsAsync(oldDocument, newDocument, cancellationToken).ConfigureAwait(false);
+        }
 
-            if (oldProject == null)
+        public static async ValueTask<bool> HasChangesAsync(Solution oldSolution, Solution newSolution, CancellationToken cancellationToken)
+        {
+            if (oldSolution == newSolution)
             {
-                EditAndContinueWorkspaceService.Log.Write("EnC state of '{0}' [0x{1:X8}] queried: project not loaded", newProject.Id.DebugName, newProject.Id);
+                return false;
+            }
 
-                // TODO (https://github.com/dotnet/roslyn/issues/1204):
-                //
-                // When debugging session is started some projects might not have been loaded to the workspace yet (may be explicitly unloaded by the user).
-                // We capture the base solution. Edits in files that are in projects that haven't been loaded won't be applied
-                // and will result in source mismatch when the user steps into them.
-                //
-                // We can allow project to be added by including all its documents here.
-                // When we analyze these documents later on we'll check if they match the PDB.
-                // If so we can add them to the committed solution and detect further changes.
-                // It might be more efficient though to track added projects separately.
+            foreach (var newProject in newSolution.Projects)
+            {
+                var oldProject = oldSolution.GetProject(newProject.Id);
+                if (oldProject == null || await HasChangedOrAddedDocumentsAsync(oldProject, newProject, changedOrAddedDocuments: null, cancellationToken).ConfigureAwait(false))
+                {
+                    return true;
+                }
+            }
 
-                return;
+            foreach (var oldProject in oldSolution.Projects)
```

Comment:
Correct me if I'm wrong, but if we got past the previous check, so we know there are no new projects, we can therefore just check if `oldSolution.Projects.Count == newSolution.Projects.Count`, right?

Reply:
Yup. Will simplify.